File: multi_agent/projects/im-backend/src/main/java/com/im/backend/service/MessageDraftService.java
package com.im.backend.service;

import com.im.backend.entity.MessageDraft;
import com.im.backend.repository.MessageDraftRepository;
import com.im.backend.dto.DraftRequest;
import com.im.backend.dto.DraftResponse;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class MessageDraftService {

    private final MessageDraftRepository draftRepository;

    @Transactional
    public DraftResponse saveDraft(DraftRequest request) {
        Optional<MessageDraft> existing = draftRepository.findByUserIdAndConversationIdAndIsDeletedFalse(
                request.getUserId(), request.getConversationId());

        MessageDraft draft;
        if (existing.isPresent()) {
            draft = existing.get();
            draft.setContent(request.getContent());
            draft.setMentionIds(request.getMentionIds());
            draft.setReplyMessageId(request.getReplyMessageId());
            draft.setMessageType(request.getMessageType());
            draft.setUpdatedAt(System.currentTimeMillis());
        } else {
            draft = MessageDraft.builder()
                    .userId(request.getUserId())
                    .conversationId(request.getConversationId())
                    .content(request.getContent())
                    .mentionIds(request.getMentionIds())
                    .replyMessageId(request.getReplyMessageId())
                    .messageType(request.getMessageType())
                    .updatedAt(System.currentTimeMillis())
                    .isDeleted(false)
                    .build();
        }

        MessageDraft saved = draftRepository.save(draft);
        log.info("Draft saved for user {} in conversation {}", request.getUserId(), request.getConversationId());
        return toResponse(saved);
    }

    @Transactional
    public void deleteDraft(Long userId, String conversationId) {
        draftRepository.deleteByUserIdAndConversationId(userId, conversationId);
        log.info("Draft deleted for user {} in conversation {}", userId, conversationId);
    }

    public Optional<DraftResponse> getDraft(Long userId, String conversationId) {
        return draftRepository.findByUserIdAndConversationIdAndIsDeletedFalse(userId, conversationId)
                .map(this::toResponse);
    }

    public List<DraftResponse> getAllDrafts(Long userId) {
        return draftRepository.findByUserIdAndIsDeletedFalseOrderByUpdatedAtDesc(userId)
                .stream().map(this::toResponse).collect(Collectors.toList());
    }

    @Transactional
    public void syncDraftToDevice(Long userId, String conversationId, Long deviceId) {
        Optional<MessageDraft> draft = draftRepository.findByUserIdAndConversationIdAndIsDeletedFalse(userId, conversationId);
        if (draft.isPresent()) {
            log.info("Syncing draft to device {} for user {} in conversation {}", deviceId, userId, conversationId);
        }
    }

    private DraftResponse toResponse(MessageDraft draft) {
        return DraftResponse.builder()
                .id(draft.getId())
                .userId(draft.getUserId())
                .conversationId(draft.getConversationId())
                .content(draft.getContent())
                .mentionIds(draft.getMentionIds())
                .replyMessageId(draft.getReplyMessageId())
                .messageType(draft.getMessageType())
                .updatedAt(draft.getUpdatedAt())
                .build();
    }
}

File: multi_agent/projects/im-desktop/src/services/message-draft-service.ts
// Message Draft Service - handles draft save/delete/sync across devices
import axios from 'axios';
import { MessageDraft, DraftRequest } from '../types/message-draft';

const API_BASE = 'http://localhost:8080/api/draft';

export class MessageDraftService {
  private userId: number;

  constructor(userId: number) {
    this.userId = userId;
  }

  async saveDraft(conversationId: string, content: string, options?: {
    mentionIds?: string;
    replyMessageId?: string;
    messageType?: string;
  }): Promise<MessageDraft> {
    const request: DraftRequest = {
      userId: this.userId,
      conversationId,
      content,
      mentionIds: options?.mentionIds,
      replyMessageId: options?.replyMessageId,
      messageType: options?.messageType,
    };
    const response = await axios.post(`${API_BASE}/save`, request);
    return response.data;
  }

  async deleteDraft(conversationId: string): Promise<void> {
    await axios.delete(`${API_BASE}/delete`, {
      params: { userId: this.userId, conversationId },
    });
  }

  async getDraft(conversationId: string): Promise<MessageDraft | null> {
    try {
      const response = await axios.get(`${API_BASE}/get`, {
        params: { userId: this.userId, conversationId },
      });
      return response.data;
    } catch (e: any) {
      if (e.response?.status === 204) return null;
      throw e;
    }
  }

  async getAllDrafts(): Promise<MessageDraft[]> {
    const response = await axios.get(`${API_BASE}/list`, {
      params: { userId: this.userId },
    });
    return response.data;
  }

  async syncDraftToDevice(conversationId: string, deviceId: number): Promise<void> {
    await axios.post(`${API_BASE}/sync`, null, {
      params: { userId: this.userId, conversationId, deviceId },
    });
  }
}
